### ISSUE TITLE
failure assigning differing arguments to `Iterable[T]` arguments

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#3055]
Self-contained repro with a user-defined function:

```py
from collections.abc import Iterable

def g[T](a: Iterable[T], b: Iterable[T]) -> None:
    pass

def f(
    a: tuple[type[int], str] | tuple[type[str], str],
    b: tuple[str, str],
) -> None:
    g(a, b)  # Expected `Iterable[str]`, found `tuple[type[int], str] | tuple[type[str], str]`
```

ty seems to fix `T` based on the `tuple[str, str]` type (this occurs no matter which order `a` and `b` are passed in), and error on the type of `a`, rather than solving `T` to a type that can accommodate all arguments.

This also repros if the formal parameter is `*iterables: Iterable[T]` rather than two separate parameters, and/or if the actual argument is `*` unpacked from a tuple rather than separate arguments.

Repro as originally observed with `itertools.chain` constructor:

```py
import itertools

def f(
    p: tuple[
        tuple[type[int], str] | tuple[type[str], str],
        tuple[str, str],
    ],
) -> None:
    x = itertools.chain(*p)
```

**Comment by sharifhsn** [comment]
Closed by https://github.com/astral-sh/ruff/pull/24837

Repro:

```py
from collections.abc import Iterable

def g[T](a: Iterable[T], b: Iterable[T]) -> None:
    pass

def f(
    a: tuple[type[int], str] | tuple[type[str], str],
    b: tuple[str, str],
) -> None:
    g(a, b)
```

ty 0.0.32:

```
error[invalid-argument-type]: Argument to function `g` is incorrect
  --> 3055.py:10:7
   |
10 |     g(a, b)
   |       ^ Expected `Iterable[str]`, found `tuple[type[int], str] | tuple[type[str], str]`

Found 1 diagnostic
```

ty 0.0.33:

```
All checks passed!
```

**Comment by charliermarsh** [comment]
(\cc @carljm in case you disagree with closing, since that PR was for the old constraint solver)

**Comment by carljm** [comment]
@charliermarsh no, not at all -- if it works in ty today, we should close it. If we regress some behavior in a future PR switching to the new constraint solver, we can open an issue then.

The one thing I do wonder about some of these old issues we are closing is whether we should be adding tests from them, so we notice if it regresses in future. Some may already be effectively tested by the PRs that close them...
